kaggle_environments/envs/lux_ai_2022/luxai2022/env.py

```diff
@@ -0,0 +1,618 @@
+                elif self.state.board.ore[unit.pos.y, unit.pos.x] > 0:
+                    unit.add_resource(1, unit.unit_cfg.DIG_RESOURCE_GAIN)
+                unit.power -= math.ceil(self.state.env_cfg.ROBOTS[unit.unit_type.name].DIG_COST * weather_cfg["power_loss_factor"])
+
+            for unit, self_destruct_action in actions_by_type["self_destruct"]:
+                unit: Unit
+                self_destruct_action: SelfDestructAction
+                pos_hash = self.state.board.pos_hash(unit.pos)
+                del self.state.board.units_map[pos_hash]
+                self.destroy_unit(unit)
+
+            for factory, factory_build_action in actions_by_type["factory_build"]:
+                factory: Factory
+                factory_build_action: FactoryBuildAction
+                team = self.state.teams[factory.team.agent]
+                self.add_unit(
+                    team=team,
+                    unit_type=factory_build_action.unit_type,
+                    pos=factory.pos.pos,
+                )
+                if factory_build_action.unit_type == UnitType.HEAVY:
+                    factory.sub_resource(3, self.env_cfg.ROBOTS["HEAVY"].METAL_COST)
+                    factory.sub_resource(4, math.ceil(self.env_cfg.ROBOTS["HEAVY"].POWER_COST * weather_cfg["power_loss_factor"]))
+                else:
+                    factory.sub_resource(3, self.env_cfg.ROBOTS["LIGHT"].METAL_COST)
+                    factory.sub_resource(4, math.ceil(self.env_cfg.ROBOTS["LIGHT"].POWER_COST * weather_cfg["power_loss_factor"]))
+
+            new_units_map: Dict[str, List[Unit]] = defaultdict(list)
+            heavy_entered_pos: Dict[str, List[Unit]] = defaultdict(list)
+            light_entered_pos: Dict[str, List[Unit]] = defaultdict(list)
+
+            for unit, move_action in actions_by_type["move"]:
+                move_action: MoveAction
+                # skip move center
+                if move_action.move_dir == 0:
+                    continue
+                old_pos_hash = self.state.board.pos_hash(unit.pos)
+                target_pos = unit.pos + move_action.dist * move_deltas[move_action.move_dir]
+                rubble = self.state.board.rubble[target_pos.y, target_pos.x]
+                power_required = unit.unit_cfg.MOVE_COST + unit.unit_cfg.RUBBLE_MOVEMENT_COST * rubble
+                power_required = math.ceil(power_required * weather_cfg["power_loss_factor"])
+                unit.pos = target_pos
+                new_pos_hash = self.state.board.pos_hash(unit.pos)
+
+                # remove unit from map temporarily
+                if len(self.state.board.units_map[old_pos_hash]) == 1:
+                    del self.state.board.units_map[old_pos_hash]
+                else:
+                    self.state.board.units_map[old_pos_hash].remove(unit)
+
+                new_units_map[new_pos_hash].append(unit)
+                unit.power -= power_required
```

Comment:
if we don't validate that they have enough power to move, this could put them at negative power

Reply:
There's a whole "action validation " function so all actions here are assumed to be valid